File: Solvers/mosek/10.0/tools/examples/fusion/python/pinfeas.py
##
#   Copyright: Copyright (c) MOSEK ApS, Denmark. All rights reserved.
#
#   File:      pinfeas.py
#
#   Purpose: Demonstrates how to fetch a primal infeasibility certificate
#            for a linear problem
#
##

from mosek.fusion import *
import sys

# Analyzes and prints infeasibility certificate for a single object,
# which can be a variable or constraint
def analyzeCertificate(name,          # name of the analyzed object
                       size,          # size of the object
                       duals,         # double[], actual dual values
                       eps):          # tolerance determining when a dual value is considered important
    for i in range(0, size):
        if abs(duals[i]) > eps:
            print(f"{name}[{i}], dual = {duals[i]}")

# Construct the sample model from the example in the manual
sMat = Matrix.sparse(3, 7, [0,0,1,1,2,2,2],
                           [0,1,2,3,4,5,6],
                           [1.0]*7)
sBound = [200, 1000, 1000]
dMat = Matrix.sparse(4, 7, [0,0,1,2,2,3,3],
                           [0,4,1,2,5,3,6],
                           [1.0]*7)
dBound = [1100, 200, 500, 500]
c = [1, 2, 5, 2, 1, 2, 1]

M = Model("pinfeas")
x = M.variable("x", 7, Domain.greaterThan(0))
s = M.constraint("s", Expr.mul(sMat, x), Domain.lessThan(sBound))
d = M.constraint("d", Expr.mul(dMat, x), Domain.equalsTo(dBound))
M.objective(ObjectiveSense.Minimize, Expr.dot(c,x))

# Set useful debugging tools and solve the model
M.setLogHandler(sys.stdout)
M.writeTask("pinfeas.ptf")
M.solve()

# Check problem status
if M.getProblemStatus() == ProblemStatus.PrimalInfeasible:
    # Set the tolerance at which we consider a dual value as essential
    eps = 1e-7

    # We want to retrieve infeasibility certificates
    M.acceptedSolutionStatus(AccSolutionStatus.Certificate)

    # Go through variable bounds
    print("Variable bounds important for infeasibility: ")
    analyzeCertificate(name = "x", size = x.getSize(), duals = x.dual(), eps = eps)

    # Go through constraint bounds
    print("Constraint bounds important for infeasibility: ")
    analyzeCertificate(name = "s", size = s.getSize(), duals = s.dual(), eps = eps)
    analyzeCertificate(name = "d", size = d.getSize(), duals = d.dual(), eps = eps)

else:
    print("The problem is not primal infeasible, no certificate to show")

File: Solvers/mosek/10.0/tools/examples/fusion/dotnet/pinfeas.cs
////
//   Copyright: Copyright (c) MOSEK ApS, Denmark. All rights reserved.
//
//   File:      pinfeas.cs
//
//   Purpose: Demonstrates how to fetch a primal infeasibility certificate
//            for a linear problem
////
using System;
using mosek.fusion;

namespace mosek.fusion.example
{
  public class pinfeas {
    //Analyzes and prints infeasibility certificate for a single object,
    //which can be a variable or constraint
    public static void analyzeCertificate(String name,          // name of the analyzed object
                                          long size,            // size of the object
                                          double[] duals,       // actual dual values
                                          double eps) {         // tolerance determining when a dual value is considered important
      for(int i = 0; i < size; i++) {
        if (Math.Abs(duals[i]) > eps) 
          Console.WriteLine("{0}[{1}],  dual = {2}", name, i, duals[i]);
      }
    }

    public static void Main(string[] args) {
      // Construct the sample model from the example in the manual
      Matrix sMat = Matrix.Sparse(3, 7, new int[]{0,0,1,1,2,2,2},
                                        new int[]{0,1,2,3,4,5,6},
                                        new double[]{1,1,1,1,1,1,1});
      double[] sBound = new double[]{200, 1000, 1000};
      Matrix dMat = Matrix.Sparse(4, 7, new int[]{0,0,1,2,2,3,3},
                                        new int[]{0,4,1,2,5,3,6},
                                        new double[]{1,1,1,1,1,1,1});
      double[] dBound = new double[]{1100, 200, 500, 500};
      double[] c = new double[]{1, 2, 5, 2, 1, 2, 1};

      using(Model M = new Model("pinfeas"))
      {
        Variable x = M.Variable("x", 7, Domain.GreaterThan(0));
        Constraint s = M.Constraint("s", Expr.Mul(sMat, x), Domain.LessThan(sBound));
        Constraint d = M.Constraint("d", Expr.Mul(dMat, x), Domain.EqualsTo(dBound));
        M.Objective(ObjectiveSense.Minimize, Expr.Dot(c,x));

        // Useful for debugging
        M.WriteTask("pinfeas.ptf");     // Save to a readable file
        M.SetLogHandler(Console.Out);   // Enable log output

        // Solve the problem
        M.Solve();

        // Check problem status
        if (M.GetProblemStatus() == ProblemStatus.PrimalInfeasible) {
          // Set the tolerance at which we consider a dual value as essential
          double eps = 1e-7;

          // We want to retrieve infeasibility certificates
          M.AcceptedSolutionStatus(AccSolutionStatus.Certificate);

          // Go through variable bounds
          Console.WriteLine("Variable bounds important for infeasibility: ");
          analyzeCertificate("x", x.GetSize(), x.Dual(), eps);

          // Go through constraint bounds
          Console.WriteLine("Constraint bounds important for infeasibility: ");
          analyzeCertificate("s", s.GetSize(), s.Dual(), eps);
          analyzeCertificate("d", d.GetSize(), d.Dual(), eps);
        }
        else {
          Console.WriteLine("The problem is not primal infeasible, no certificate to show");
        }
      }
    }
  }
}
